docs: add hydration guide (#49665)
angular/angular: aio/content/guide/hydration.md

# Hydration

<div class="alert is-important">

Hydration feature is available for [developer preview](https://angular.io/guide/releases#developer-preview).
It's ready for you to try, but it might change before it is stable.

</div>

## What is hydration

Hydration is the process that restores the server side rendered application on the client. This includes things like reusing the server rendered DOM structures, persisting the application state, transferring application data that was retrieved already by the server, and other processes.

## Why is hydration important?

> Hydration improves application performance by avoiding extra work to re-create DOM nodes. Instead, Angular tries to match existing DOM elements to the applications structure at runtime and reuses DOM nodes when possible. This results in a performance improvement that can be measured using Core Web Vitals (CWV) statistics, such as reducing the First Input Delay (FID) and Largest Contentful Paint (LCP), as well as Cumulative Layout Shift (CLS). Improving these numbers also affects things like SEO performance.

Reviewer: Nit: consider linking to the web.dev definitions for LCP, CLS, etc, as some folks may not be familiar (e.g. https://web.dev/lcp/)

> Without hydration enabled, server side rendered applications may see a flicker as the page initially loads due to that re-rendering that occurs. Angular will destroy and re-render the application, which will also negatively impact all of the Core Web Vitals mentioned. The re-rendering means an additional delay for every one of the statistics, and a potential for layout shift with all of the imagery. Hydration prevents that flickering from happening.

Reviewer: Nit: would suggest softening the language around expected CWV diff, since it will depend on the case, eg:

```suggestion
Without hydration enabled, Angular will destroy and re-render the application's DOM, which may result in a visible UI flicker. This re-rendering can negatively impact Core Web Vitals like LCP and cause a layout shift. Enabling hydration allows the existing DOM to be re-used and prevents a flicker.
```

```typescript
import {
  bootstrapApplication,
  provideClientHydration,
} from '@angular/platform-browser';
...

bootstrapApplication(RootCmp, {
  providers: [provideClientHydration()]
});
```

Alternatively if you are using NgModules, you would add `provideClientHydration` to your root app module's provider list.

```typescript
import {provideClientHydration} from '@angular/platform-browser';
import {NgModule} from '@angular/core';

@NgModule({
  declarations: [RootCmp],
  exports: [RootCmp],
  bootstrap: [RootCmp],
  providers: [provideClientHydration()],
})
export class AppModule {}
```

After you've followed these steps and have started up your server, load your application in the browser. You can confirm hydration is enabled by opening Developer Tools in your browser and viewing the console. You should see a message that includes hydration-related stats, such as the number of components and nodes hydrated. Note: Angular calculates the stats based on all components rendered on a page, including those that come from third-party libraries.

<a id="limitations"></a>

> ## Limitations

Reviewer: Nit: "limitations" sounds a bit scary, esp since parallel DOM structure is a baseline expectation in other SSR frameworks. Would something like "Compatibility" or "Constraints" work?

Hydration imposes a few limitations on your application that are not present without hydration enabled. Ideally your application should have similar application structure (in terms of its DOM representation) on both the server and the client. The process of hydration expects the DOM tree to have the same structure in both places.

### Direct DOM Manipulation

If you have components that manipulate the DOM using native DOM APIs, the hydration process will encounter errors. Specific cases where DOM manipulations is a problem are situations like accessing the `document`, querying for specific elements, and injecting additional nodes using `appendChild`. Detaching DOM nodes and moving them to other locations will also result in errors.

This is because Angular is unaware of these DOM changes and cannot resolve them during the hydration process. Angular will expect a certain structure, but it will encounter a different structure when attempting to hydrate. This mismatch will result in hydration failure and throw a DOM mismatch error ([see below](#errors)).

> It is best to refactor your component to avoid this sort of DOM manipulation. Try to use Angular APIs to do this work, if you can. If you cannot refactor this behavior, use the `ngSkipHydration` attribute ([described below](#ngskiphydration)) until you can refactor into a hydration friendly solution.

Reviewer: Nit: It seems like fixing DOM manipulation is buried a bit in the guide. Intuitively, I'd expect this to be in the top level instructions for enabling hydration, as you're very likely to encounter this situation when migrating. Maybe it makes sense to add a final step for "replace direct DOM manipulation with Angular constructs or 'ngSkipHydration'", with a link to the limitations section below?

Author: I've added a bit of a stronger alert talking about this in the [follow up PR](https://github.com/angular/angular/pull/49705). PTAL

# Hydration

> 

Reviewer 2: I think we'd need to include a note here that this is a developer preview feature and what it means. Something like this: https://github.com/angular/angular/commit/b136e9eea6c9b22796186262292db738cd4d1a12#diff-c4936d15746d95257ed60d75f6b4dfd32fc6ed129c434de84e397a790dbe3893.

Author: SGTM. I'll add a similar note.

<div class="alert is-important">

Hydration is available for [developer preview](https://angular.io/guide/releases#developer-preview).
It's ready for you to try, but it might change before it is stable.

</div>

## What is hydration

Hydration is the process that restores the server side rendered application on the client. This includes things like reusing the server rendered DOM structures, persisting the application state, transferring application data that was retrieved already by the server, and other processes.

## Why is hydration important?

> Hydration of applications improves application performance by avoiding extra work to re-create DOM nodes. Instead, Angular tries to match existing DOM elements to the applications structure at runtime and reuses DOM

Reviewer 2: ("application" is used twice, we can drop the first occurrence)
```suggestion
Hydration improves application performance by avoiding extra work to re-create DOM nodes. Instead, Angular tries to match existing DOM elements to the applications structure at runtime and reuses DOM
```

> Without hydration enabled, server side rendered apps may see a flicker as the page initially loads due to that re-rendering that occurs. Hydration prevents that flickering from happening, negating the need for `enabledBlocking` on `InitialNavigation` in the router.

Reviewer 2: Ah, I see that we talk about this below, so may be we can restructure the content a bit to start with the user impact.

Author: I've updated this section. PTAL.

> ## How do you enable hydration in angular

Reviewer 2: ```suggestion
## How do you enable hydration in Angular
```

nodes when possible. This results in a performance improvement that can be measured using Core Web Vitals (CWV) statistics, such as reducing the First Input Delay (FID) and Largest Contentful Paint (LCP), as well as Cumulative Layout Shift (CLS). Improving these numbers also affects things like SEO performance.

> Without hydration enabled, server side rendered apps may see a flicker as the page initially loads due to that re-rendering that occurs. Angular will destroy and re-render the application, which will also negatively impact all of the Core Web Vitals mentioned. The re-rendering means an additional delay for every one of the statistics, and a potential for layout shift with all of the imagery. Hydration prevents that flickering from happening.

Reviewer 2: nit: "app" vs "application" (I think it's ok to use either, we can just check that we do that consistently in this guide)
```suggestion
Without hydration enabled, server side rendered applications may see a flicker as the page initially loads due to that re-rendering that occurs. Angular will destroy and re-render the application, which will also negatively impact all of the Core Web Vitals mentioned. The re-rendering means an additional delay for every one of the statistics, and a potential for layout shift with all of the imagery. Hydration prevents that flickering from happening.
```

Author: I've updated it to be `applications`.

> You can enable hydration by visiting your main app component or module and importing `provideClientHydration` from `@angular/platform-browser`. You'll then add that provider to your app's bootstrapping providers list.

Reviewer 2: We can probably mention here that the NgUniversal should be enabled first and refer to the Universal guide. This should be helpful to establish the context for people that come to this page directly.

Author: Added this and linked to the universal guide.

import {
  bootstrapApplication,
  provideClientHydration,
} from '@angular/platform-browser';
...

bootstrapApplication(RootCmp, {
  providers: [provideClientHydration()]
});
```

> Alternatively if you are using ngModules, you would add `provideClientHydration` to your root app module's provider list.

Reviewer 2: ```suggestion
Alternatively if you are using NgModules, you would add `provideClientHydration` to your root app module's provider list.
```

```typescript
import {provideClientHydration} from '@angular/platform-browser';
import {NgModule} from '@angular/core';

@NgModule({
  declarations: [RootCmp],
  exports: [RootCmp],
  bootstrap: [RootCmp],
  providers: [provideClientHydration()],
})

> export class RootModule {}

Reviewer 2: I think in CLI-generated apps the root module is named as `AppModule`.
```suggestion
export class AppModule {}
```

```

> After you've followed these steps and have started up your server, load your application in the browser. You can confirm hydration is enabled by opening Developer Tools and viewing the console. You should see a message in the console that has this structure:

Reviewer 2: I think we can avoid mentioning a particular message (so that it doesn't go out of sync with the code):
```suggestion
After you've followed these steps and have started up your server, load your application in the browser. You can confirm hydration is enabled by opening Developer Tools in your browser and viewing the console. You should see a message that includes hydration-related stats, such as the number of components and nodes hydrated. Note: Angular calculates the stats based on all components rendered on a page, including those that come from third-party libraries.
```

> ```text
> Angular hydrated {x} component(s) and {y} node(s), {z} component(s) were skipped. Learn more at https://angular.io/guides/hydration.
> ```
> 
> You can also look at ngDevMode object in the console. There should be three fields on the ngDevMode object that report the numbers above. You'll find them in the fields `hydratedComponents`, `hydratedNodes`, and `componentsSkippedHydration`. If hydration is enabled, you should see positive numbers in those fields, with the exception being `componentsSkippedHydration` if no components were skipped.

Reviewer 2: I'd propose dropping this, since it's too technical and the `ngDevMode` is not a public API (so we may decide to remove/rename it later).
```suggestion
```

## Limitations

> Hydration imposes a few limitations on your application that are not present without hydration enabled. Ideally your application should have similar application structure on both the server and the client. The process of hydration expects the structure to be the same in both places.

Reviewer 2: ```suggestion
Hydration imposes a few limitations on your application that are not present without hydration enabled. Ideally your application should have similar application structure (in terms of its DOM representation) on both the server and the client. The process of hydration expects the DOM tree to have the same structure in both places.
```

### Direct DOM Manipulation

> If you have components that manipulate the DOM through direct DOM manipulation using native DOM APIs, the hydration process will encounter errors. Specific cases where DOM manipulations is a problem are situations like accessing the `document`, querying for specific elements, and injecting additional nodes using `appendChild`. Detaching DOM nodes and moving them to other locations will also result in errors.

Reviewer 2: ```suggestion
If you have components that manipulate the DOM using native DOM APIs, the hydration process will encounter errors. Specific cases where DOM manipulations is a problem are situations like accessing the `document`, querying for specific elements, and injecting additional nodes using `appendChild`. Detaching DOM nodes and moving them to other locations will also result in errors.
```

This is because Angular is unaware of these DOM changes and cannot resolve them during the hydration process. Angular will expect a certain structure, but it will encounter a different structure when attempting to hydrate. This mismatch will result in hydration failure and throw a DOM mismatch error (see below).

> It is best to refactor your component to avoid this sort of DOM manipulation. Try to use Angular APIs to do this work, if you can. If you cannot refactor this behavior, use the `ngSkipHydration` attribute below until you can refactor into a hydration friendly solution.

Reviewer 2: ```suggestion
It is best to refactor your component to avoid this sort of DOM manipulation. Try to use Angular APIs to do this work, if you can. If you cannot refactor this behavior, use the `ngSkipHydration` attribute (described below) until you can refactor into a hydration friendly solution.
```

There are a few cases where if you have a component template that does not have valid HTML structure, this could result in a DOM mismatch error during hydration. For example, if you have a `<table>` and you did not add a `<tbody>` around your rows, the browser will automatically insert one for you when it sees this HTML structure. The hydration process will not know that the `<tbody>` exists and will result in a hydration mismatch error. The solution to this problem is to ensure your template has a valid `<tbody>` where it should be.

Other examples of this behavior are adding a `<div>` inside a `<p>` tag, putting a `<p>` inside of an `<a>`, and putting a `<p>` inside an `<h1>`. If you are uncertain about whether your HTML is valid, you can use a [syntax validator](https://validator.w3.org/) to check it.

## Errors

There are several hydration related errors you may encounter.

### Hydration Node Mismatch (NG500)

### Hydration Missing Siblings (NG501)
### Hydration Missing Node (NG502)
### Unsupported Projection of DOM Nodes (NG503)
### Invalid Skip Hydration Host (NG504)
* DOM Mismatch (link to the error page)
example of error
* Direct DOM manipulation (link to the error page)

> example of error

Reviewer 2: We may consider avoiding errors enumeration, so that it doesn't become stale as we add new or update existing messages/codes.

Author: Yeah, I take a look at what other guides do in regards to errors and mimic that. We can also link to the error pages, once they exist.

> ## NgSkipHydration

Reviewer 2: ```suggestion
## How to skip hydration for particular components
```

Some components may not work properly with hydration enabled due to some of the aforementioned issues, like direct DOM manipulation. As a workaround, you can add `ngSkipHydration` to a component in order to skip hydrating the entire component.

```html
<example-cmp ngSkipHydration />
```

Alternatively you can set `ngSkipHydration` as a host binding.

```typescript
@Component({
  ...
  host: {ngSkipHydration: 'true'},
})
class ExampleCmp {}
```

> `NgSkipHydration` will skip hydrating the entire component, and it's children. Using this attribute means that the component will behave as if hydration is not enabled, meaning it will destroy and re-render itself.

Reviewer 2: ```suggestion
The `ngSkipHydration` attribute will force Angular to skip hydrating the entire component, and it's children. Using this attribute means that the component will behave as if hydration is not enabled, meaning it will destroy and re-render itself.
```

<div class="alert is-helpful">
  This will fix rendering issues, but it means that for this component (and its children), you don't get the benefits of hydration. You will need to adjust your component's implementation to avoid hydration-breaking patterns (i.e. Direct DOM Manipulation) to be able to remove the skip hydration annotation.
</div>

> `NgSkipHydration` can only be used on components. You cannot add it to any random DOM node, and if you do, you'll get an Invalid Skip Hydration Host error.

Reviewer 2: ```suggestion
The `ngSkipHydration` attribute can only be used on component host nodes. Angular throws an error if this attribute is added to other nodes.
```

> Keep in mind that adding `ngSkipHydration` to your root application component would effectively disable hydration for your entire application. Be careful and thoughtful about using this attribute. It is intended as a last resort workaround. Components that break hydration should be considered bugs that need to be fixed.

Reviewer 2: > Keep in mind that adding `ngSkipHydration` to your root application component would effectively disable hydration for your entire application. 

@jessicajaniuk It should do it now (once PR https://github.com/angular/angular/pull/49675 lands) :)
